tests/Representation.Tests/TestUtils/Nibble.cs

```diff
@@ -0,0 +1,183 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+using Symbolica.Expression;
+
+namespace Symbolica.Representation.TestUtils
+{
+    internal readonly struct Nibble : IEquatable<Nibble>
```

Comment:
Right, I knew you would wonder about this. I realise it looks fairly mental.

I want to work on the bit level for these tests so I can be sure that they are actually testing the parts I found tricky to write, without doing any magic dotnet type cheating. The "reference" implementations I use here in the tests are actually pretty different to the more Z3-suitable implementations in the source. Specifically, I don't want to deal with the various numerical type promotions involved with right-shifting the built-in dotnet numerical types. It does an arithmetic right shift for signed types and a logical right shift for unsigned types, and also always promotes the shift amount to an `int`. This is all fine and makes sense, but I actually just want to move the bits around without worrying about numbers. Maybe it is just because I wrote it, but the `Nibble` type is basically built from first principles using simple boolean operators and I actually find it fairly trivial to follow (although it is horrifically inefficient).

Also, 4 bits is a nice size for the testing I actually want to do here, where I really care about the overflows etc, so I want to test the full cartesian product of everything. I end up with a few thousand tests rather than the tens of millions I would have if I did a full byte the same way. This is just an aside, I know I could simply `0x0F` mask a byte after every operation.